### PR TITLE
Add KuMir

### LINKED ISSUE #2759: Support KuMir
**Language**

“КуМир” (KuMir) is a Russian-based, ALGOL-like educational programming language. KuMir is an acronym for the “Kit of training worlds”. (In particular, in KuMir you can control a robot that moves along grids, which are called worlds.) The literal translation of “кумир” is “idol”.

In 1985 [Andrey Yershov](https://en.wikipedia.org/wiki/Andrey_Ershov) made a [series of publications](http://ershov.iis.nsk.su/ru/ershov/scient) on teaching programming in Soviet schools. In those publications, Yershov proposed a deliberately unnamed, Russian-based, ALGOL-like algorithmic language. Yershov used the language in his [textbook](http://publ.lib.ru/ARCHIVES/E/ERSHOV_Andrey_Petrovich/_Ershov_A.P..html) in the same year.

The reserved words of the language are:

- `алг` (`proc`, a procedure name), `арг` (formal argument), `все` (`esac` or `fi`, end of condition), `выбор` (`case` in `case`...`in`), `для` (`for`), `до` (`to` in `for`...`to`), `если` (`if`), `знач` (the `Result` variable in Pascal), `иначе` (`else`), `кон` (end of program), `кц` (`od`, end of loop), `нач` (`begin` of program), `не` (`not`), `нц` (`do`, begin of loop), `от` (`from` in `for`...`from`...`to`), `пока` (`while`), `при` (`in` in `case`...`in` and `until` in `for`...`until`), `рез` (`proc` in arguments, a pass-by-reference parameter for storing a result), `то` (`then` in `if`...`then`), `шаг` (`step` in `for`...`step`...`until`);
- data type names: `вещ` (real), `вещ таб` (array of real), `дроб` (fractional), `дроб таб`, `лит` (string), `лит таб`, `лог` (Boolean), `лог таб`, `нат` (natural [number]), `нат таб`, `цел` (integer), and `цел таб`.

Operators are `и` (and), `или` (or), `*`, `**` (power), `+`, `-`, `/`, `<`, `⩽` (later `<=`), `=`, `≠` (later `<>`), `>`, `⩾` (later `>=`).

Punctuation marks are `"`, `(`, `)`, `,`, `;`, `:`, `:=`, `[`, `]`.

In 1990 Yershov’s algorithmic language was used in a [textbook](http://publ.lib.ru/ARCHIVES/K/KUSHNIRENKO_Anatoliy_Georgievich/_Kushnirenko_A.G..html) by [Anatoly Koushnirenko](https://en.wikipedia.org/wiki/Bernstein%E2%80%93Kushnirenko_theorem), [Gennady Lebedev](https://en.wikipedia.org/wiki/Gennady_Lebedev), and [Rudolf Svoren](https://ru.wikipedia.org/wiki/%D0%A1%D0%B2%D0%BE%D1%80%D0%B5%D0%BD%D1%8C,_%D0%A0%D1%83%D0%B4%D0%BE%D0%BB%D1%8C%D1%84_%D0%90%D0%BD%D0%B0%D1%82%D0%BE%D0%BB%D1%8C%D0%B5%D0%B2%D0%B8%D1%87). They had

- added several reserved words: `ввод` (input), `вывод` (output), `дано` (precondition), `исп` (begin module), `надо` (postcondition), `нс` (newline), `раз` (an alias for `for`...`step 1`), `утв` (assertion);
- added `|` as a comment operator;
- added Boolean values `да` (true) and `нет` (false);
- added `'` (same as `"`);
- added `сим` (character) and `сим таб` data types;
- dropped `дроб`, `дроб таб`, `нат`, and `нат таб` data types.

Koushnirenko’s dialect is named “School algorithmic language”. In 1990–1996 InfoMir Co. created [KuMir IDE](https://web.archive.org/web/20130604044311/http://www.infomir.ru/old4/kumir.htm) for several platforms including DOS and Mac OS 7. These implementations

- support the `выход` reserved word (`break` in Pascal);
- don’t require space before `таб`.

Since mid-2000s KuMir (along with C, Basic, Pascal, and Python) has been used on the State exams in [junior (9th grade)](https://fipi.ru/oge/demoversii-specifikacii-kodifikatory#!/tab/173801626-5) and [senior (11th grade)](https://fipi.ru/ege/demoversii-specifikacii-kodifikatory#!/tab/151883967-5) middle schools in Russia. Although called [Yershov’s/Koushnirenko’s] algorithmic language (“алгоритмический язык”), it is definitely KuMir, because it doesn’t have spaces before `таб`. See the [2021 exam specification](http://doc.fipi.ru/gve/gve-11/2021/spec_INF_pism_gve-11_2021.pdf) (pages 18, 21, 24, 25).

[NIISI RAS](https://en.wikipedia.org/wiki/Scientific_Research_Institute_of_System_Development) supports a [Qt-based IDE for KuMir](https://www.niisi.ru/kumir/). The IDE is included in “[Альт Образование](https://www.basealt.ru/fileadmin/user_upload/polifrafya/Application_Software.pdf)”, a Linux distribution for middle and vocational schools.

NIISI RAS

- added `аргрез` (an alias for `арг рез`), `ВКЛЮЧИТЬ` (include), `всё` (an alias for `все`), `дс`, `использовать` (use), `кон_исп` (an alias for `кон исп`), `кц_при` (an alias for `кц при`), `пауза` (breakpoint), `стоп` (`exit` in Pascal);
- added data type names: `компл` (complex number), `сканкод` (scan code), `файл` (file), `цвет` (color).

_**General rules**_

KuMir is case-sensitive.

Source code file names have the extension `kum`.

Since this is an educational language, it is desirable to retain the ability to use the default color scheme:

- comments should be highlighted in gray;
- constants (numbers, strings and Boolean values) should be highlighted in light-blue;
- data type names should be highlighted in brown;
- names should be highlighted in blue.

_**Names**_

Identifiers can consist of digits, `@`, `_`, Latin and Cyrillic letters of any case, `!` (undocumented), and spaces (`/\s/`) except `\n` and `\r`. Leading and trailing spaces are truncated. Adjacent inner spaces are merged. Identifiers with and without inner spaces are not equivalent. I. e., `a␣b`, `a␣␣b`, and `a␣␣␣b` are equivalent, whereas `a␣b` and `ab` are not (␣ represents a single space).

In fact, the modern KuMir compiler allows absolutely any character inside an identifier, besides `[\x00-\x1f\x22-\x2f\x3a-\x3f\x5b-\x5e\x60\x7b-\x7e]`. The first character can be represented as `[^\x00-\x20\x22-\x3f\x5b-\x5e\x60\x7b-\x7e]`, because an identifier cannot begin with a numeric digit. The subsequent part can be represented as `[^\x00-\x1f\x22-\x2f\x3a-\x3f\x5b-\x5e\x60\x7b-\x7e]+`.

It is impossible to distinguish a non-parameterized function call from a variable name.

A single `не` keyword can interrupt a Boolean _variable_ name on non-last position. For example, `не a b c`, `a не b c`, `a b не c` is equivalent to the inversion of the `a b c` variable.

_**Strings**_

Strings must be surrounded by single (`'`) or double (`"`) quotes. The quotes are not interchangeable.

No escaping and no interpolation allowed. Newlines (`\n` and `\r`; `U+2028` and `U+2029` are not newlines) are prohibited inside strings. If you want to insert a newline into a string, concatenate the string with the `нс` constant.

_**Data type names**_

`вещ`, `лит`, `лог`, `сим`, `цел` are common data type names. You can combine these reserved words with the `таб` reserved word to make an array. Only `[\x20]*` are allowed between a type name and `таб`, i. e. `вещтаб`, `вещ␣таб`, `вещ␣␣таб`, etc.

`компл`, `сканкод`, `файл`, `цвет` are data types, that are defined in standard modules (aka executors). You can’t combine them with `таб`. These words are not reserved until you import the corresponding module.

_**Numeric constants**_

The only difference from C-like numbers is the prefix indicating a hexadecimal number (`$`). We should search for names before searching for numbers, because names can contain digits surrounded by spaces.

**Additional resources**

- [KuMir IDE](https://www.niisi.ru/kumir/dl.htm) by NIISI RAS ([source code](https://git.niisi.ru/kumir/kumir2));
- a [bibliography](https://kpolyakov.spb.ru/school/kumir.htm) by K. Polyakov, D. Sc.

### PR DESCRIPTION
**Pull request by edukisto**
Resolves #2759.

### CODE REVIEW

**Comment by RunDevelopment** [comment]
Oh oh. Sorry @edukisto. It's my fault CI fails. I'll try to fix it immediately.

**Comment by RunDevelopment** [comment]
@edukisto I fixed it now. Please rebase and then the CI should pass. I'm sorry for the inconvenience.

**Review by RunDevelopment — CHANGES_REQUESTED** [review]
Excellent work @edukisto!

I only have a few nits and then we can merge this.

I also thoroughly enjoyed reading through #2759. Thank you for the detailed explanation of a language that otherwise has little English documentation.

File: components/prism-kumir.js [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+		},
+
+		'prolog': {
+			pattern: /#.*/,
+			greedy: true
+		},
+
+		'string': {
+			pattern: /"[^\n\r"]*"|'[^\n\r']*'/,
+			// pattern: /(["'])[^\\1\n\r]*?\1/, // 5-10% slower!
```
[the remark this answers is not in the document]
Comment (edukisto) [reply components/prism-kumir.js:31]: Thank you!

The saddest thing is that I’ve tested three correct templates, but haven’t fix the comment.

Both `/(["'])(?:(?!\1)[^\n\r])*\1/` and `/(["'])[^\n\r]*?\1/` are slower than `/"[^\n\r"]*"|'[^\n\r']*'/`.

File: components/prism-kumir.js [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+		'type': [
+			{
+				pattern: wrapId(/(^|[<nonId>])(?:вещ|лит|лог|сим|цел)(?:\x20*таб)?(?=[<nonId>]|$)/.source),
+				lookbehind: true,
+				alias: 'builtin'
+			},
+			{
+				pattern: wrapId(/(^|[<nonId>])(?:компл|сканкод|файл|цвет)(?=[<nonId>]|$)/.source),
+				lookbehind: true,
+				alias: 'inserted'
```
[the remark this answers is not in the document]
Comment (edukisto) [reply components/prism-kumir.js]: Maybe, `important`?

File: components/prism-kumir.js [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+	/**
+	 * Regular expression for characters that are not allowed in identifiers.
+	 * @type {String}
+	 */
+	var nonId = /\s\x00-\x1f\x22-\x2f\x3a-\x3f\x5b-\x5e\x60\x7b-\x7e/.source;
+
+	/**
+	 * Surround a regular expression for IDs with patterns for non-ID sequences.
+	 * @param {String} pattern A regular expression for identifiers.
+	 * @param {String} flags The regular expression flags.
```
[the remark this answers is not in the document]
Comment (edukisto) [reply components/prism-kumir.js]: Thanks!

**Review by RunDevelopment — APPROVED** [review]

**Comment by RunDevelopment** [comment]
Thank you very much for contributing @edukisto!

**Pull request merged**